### PR TITLE
Widget picker updates niteshchetry

### PR DESCRIPTION
**Pull request by niteshchetry**
Here are current screenshots of the widget picker. Still likely need to discuss UI with Laura after the 9/28/19 meeting.

Screen Shot 2019-09-27 at 6 57 29 PM
Screen Shot 2019-09-27 at 6 58 19 PM

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -22,10 +22,10 @@ import "./App.css";
 window.addEventListener("unload", function () { });
 
 type AppProps = {
-	IEOREDGE: boolean,
-	currYear: number,
-	firebase?: any,
-	ContentEditor: React.FC<ContentEditorProps>
+    IEOREDGE: boolean,
```
Comment (kwokwilliam) [inline src/App.tsx]: Can you set it so the auto format uses a tab instead of four spaces? It would also get rid of these unnecessary changes in the commit.

File: src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx [hunk 2]
```diff
@@ -50,7 +46,7 @@ export const WidgetEditor: React.FC<WidgetEditorProps> = ({ content, contentHash
     if (!content || !firebase) {
         return <></>;
     }
-
+    console.log(editedContent.type);
```
Comment (kwokwilliam) [inline src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx]: Remove console log

File: src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx [hunk 3]
```diff
@@ -64,20 +60,36 @@ export const WidgetEditor: React.FC<WidgetEditorProps> = ({ content, contentHash
...
-                            setEditedContentOnChange("type", e.target.value as string, editedContent, setEditedContent);
-                        }}>
-                        {Object.keys(WidgetTypes).map(widgetType => {
-                            return <MenuItem key={widgetType} value={widgetType}>
-                                {widgetType}
-                            </MenuItem>
-                        })}
-                    </Select>
-                </FormControl>
+                <div className="wiki-form">
```
Comment (kwokwilliam) [inline src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx]: Maybe `widget-picker` instead of `wiki-form`?

File: src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx [hunk 4]
```diff
@@ -64,20 +60,36 @@ export const WidgetEditor: React.FC<WidgetEditorProps> = ({ content, contentHash
...
+                    <form>
+                        <fieldset>
+                            <legend>Select a Widget</legend>
+                            <select id="widget" className="widget-type"
```
Comment (kwokwilliam) [inline src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx]: using the id `widget` is a bit general. I would avoid that and instead have it be something like `widget-picker-select` or something.

File: src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx [hunk 5]
```diff
@@ -64,20 +60,36 @@ export const WidgetEditor: React.FC<WidgetEditorProps> = ({ content, contentHash
...
+                                value={editedContent.type}
+                                onChange={(e) => {
+                                    setEditedContentOnChange("type", e.target.value as string, editedContent, setEditedContent);
+                                }}>
+                                <option value="" disabled selected>-- Select a Widget --</option>
+                                {/* display Widget Categories and Widgets in dropdown, sorted alphabetically */}
+                                {Object.keys(WidgetCategories).map((category) => {
```
Comment (kwokwilliam) [inline src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx]: You are mapping the widget categories (O(n)), then filtering (O(n))+ sorting (O(m)) +  mapping again (O(m)) (where m is a subset of n) inside of that. so this is roughly a O(n^2 + 2nm) ~= O(n^2) operation. Is there a way we can shorten this to only have it map stuff once, or at least not do any calculations multiple times for each widget? If it loads it's fine, but this should be put on a backlog as a high priority thing to optimize.

**Review by kwokwilliam — CHANGES_REQUESTED** [review]

File: src/App.tsx (hunk 1 above)
Comment (niteshchetry) [reply src/App.tsx]: Changed vscode settings to use indentations instead of spaces, then saved the file.

File: src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx (hunk 2 above)
Comment (niteshchetry) [reply src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx]: Removed line

File: src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx (hunk 3 above)
Comment (niteshchetry) [reply src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx]: Switched className to widget-picker and updated WidgetEditor.css accordingly

File: src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx (hunk 4 above)
Comment (niteshchetry) [reply src/components/ContentEditor/WidgetEditor/WidgetEditor.tsx]: removed both id and className for the select tag because they were actually not being called in WidgetEditor.css
